### PR TITLE
upgrade kubebuilder to plugin/v4

### PR DESCRIPTION
**Pull request by randytqwjp**
#### What this PR does / why we need it:
In place upgrade for kubebuilder to plugin v4

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

**Comment by randytqwjp** [comment]
Tests currently failing will take a look

**Comment by randytqwjp** [comment]
@sanposhiho After upgrading kubebuilder, controller with name tortoise already exists. Controller names must be unique to avoid multiple controllers reporting to the same metric. This error appears for every controller test after the first test. The way the controller tests are written for tortoise is very different from the documentation so i'm not sure how to fix this.. do you know if its possible to delete this controller in the cleanup?

**Comment by sanposhiho** [comment]
What if you move [startController()](https://github.com/mercari/tortoise/blob/df7b7b87a987363f06ffd2cbb95c478d952fec28/internal/controller/tortoise_controller_test.go#L370)  like this?

```diff
	runTest := func(path string) {
		initializeResourcesFromFiles(ctx, k8sClient, filepath.Join(path, "before"))
-		stopFunc = startController(ctx)
		if os.Getenv("UPDATE_TESTCASES") == "true" {
			generateTestCases(path)
		} else {
			checkWithWantedResources(path)
		}
		cleanUp()
	}
+	BeforeEach(func() {
+		stopFunc = startController(ctx)
+	})
```

**Comment by randytqwjp** [comment]
@sanposhiho I fixed it with this `m, err := manager.New(cfg, manager.Options{
				Controller: config.Controller{
					SkipNameValidation: ptr.To(true),
				},
			})` apparently theres a new option to skip name validation. Maybe this was not present in previous controller-runtime version

**Comment by randytqwjp** [comment]
@sanposhiho Not sure what this exec error is for the tortoisectl test would you have any clue?

File: cmd/tortoisectl/main.go [hunk 1]
```diff
@@ -1,7 +1,7 @@
-package main
+package tortoisectl
 
 import "github.com/mercari/tortoise/cmd/tortoisectl/commands"
 
-func main() {
+func tortoisectl() {
```
Comment (sanposhiho) [inline cmd/tortoisectl/main.go:5]: why renamed like this? It might be the reason

**Comment by randytqwjp** [comment]
@sanposhiho all the tests are passing but somehow lint is getting killed on the runner.. likely to be OOM? seems like the linter is using a lot of memory on the runner compared to my local

**Comment by randytqwjp** [comment]
> @sanposhiho all the tests are passing but somehow lint is getting killed on the runner.. likely to be OOM? seems like the linter is using a lot of memory on the runner compared to my local

Looks like go version issue. Specified go1.22.6 and it works now

File: test/e2e/e2e_test.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
+/*
```
Comment (sanposhiho) [inline test/e2e/e2e_test.go:1]: License is wrong.
https://github.com/mercari/tortoise/blob/main/LICENSE

File: test/e2e/e2e_suite_test.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
+/*
```
Comment (sanposhiho) [inline test/e2e/e2e_suite_test.go:1]: ditto

File: test/utils/utils.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
+/*
```
Comment (sanposhiho) [inline test/utils/utils.go:1]: ditto

**Comment by randytqwjp** [comment]
@sanposhiho could you review this PR when you have time

File: cmd/main.go [hunk 5]
```diff
@@ -120,8 +120,6 @@ func main() {
 
 	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
 		Scheme:                 scheme,
-		MetricsBindAddress:     metricsAddr,
```
[the remark this answers is not in the document]
Comment (randytqwjp) [reply cmd/main.go:123]: https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.19.1/pkg/metrics/server#Options metric server defaults to 8080 now which seems to be what you specified in the original metricsAddr `flag.StringVar(&metricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")`

File: cmd/tortoisectl/test/tortoisectl_test.go [hunk 6]
```diff
@@ -73,6 +73,10 @@ func Test_TortoiseCtlStop(t *testing.T) {
 
 	// Build the latest binary.
 	buildTortoiseCtl(t)
+	_, _, err := execCommand(t, "chmod", "+x", "./testdata/bin/tortoisectl")
```
[the remark this answers is not in the document]
Comment (randytqwjp) [reply cmd/tortoisectl/test/tortoisectl_test.go:76]: go vet fails without this line

[the remark this answers is not in the document]
Comment (sanposhiho) [reply cmd/tortoisectl/test/tortoisectl_test.go:76]: How did it fail? I don't think `go vet` blame anything even if we don't do `chmod`. The binary made from `go build` should already have `+x`.

**Review by sanposhiho — APPROVED** [review]

File: cmd/tortoisectl/test/tortoisectl_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (randytqwjp) [reply cmd/tortoisectl/test/tortoisectl_test.go]: @sanposhiho i forgot to remove the err handling for the execCommand.. Could you review again?

**Review by sanposhiho — APPROVED** [review]

**Pull request merged**